File: Top_100_Liked_Questions/138._Copy_List_with_Random_Pointer/solution.js
/**
 * // Definition for a Node.
 * function Node(val, next, random) {
 *    this.val = val;
 *    this.next = next;
 *    this.random = random;
 * };
 */

/**
 * @param {Node} head
 * @return {Node}
 */
var copyRandomList = function(head) {
    if (!head) {
        return null;
    }
    
    // create copy node and put it between its next node
    let curr = head;
    while (curr) {
        let copyNode = new Node(curr.val, curr.next, null);
        let next = curr.next;
        curr.next = copyNode;
        curr = next;
    }
    
    // set copy node's random as prev.random.next
    let prev = head;
    while (prev) {
        curr = prev.next;
        if (prev.random === null) {
            curr.random === null;
        } else {
            curr.random = prev.random.next;
        }
        prev = curr.next;
    }
    
    // separate the copy nodes and return its head
    let copyHead = head.next;
    let originalNode = head;
    let copyNode = copyHead;
    while (copyNode.next) {
        originalNode.next = copyNode.next;
        originalNode = originalNode.next;
        copyNode.next = originalNode.next;
        copyNode = copyNode.next;
    }
    originalNode.next = null;
    
    return copyHead;
};

File: Top_100_Liked_Questions/138._Copy_List_with_Random_Pointer/solution.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Node') -> 'Node':
        if not head:
            return None
        
        # make a copy node and place it between current and next node
        curr = head
        while curr:
            newNode = Node(curr.val, curr.next)
            curr.next = newNode
            curr = newNode.next
        
        # set the random of the copy nodes as prev node's random's next
        prev = head
        while prev:
            curr = prev.next
            curr.random = prev.random.next if prev.random else None
            prev = curr.next
        
        # separate the copy nodes from the list and return its head
        orig = head
        copy = copyHead = head.next
        while copy.next:
            orig.next = copy.next
            orig = copy.next
            copy.next = orig.next
            copy = orig.next
        
        orig.next = None
        
        return copyHead